expo/expo
# [expo-development-client] Add 'expo-development-client' package, add to 'bare-expo' behind feature flag

# Why

Land the 'expo-development-client' package from the development client project to master, so that it's easier to collaborate with @tsapeta and @lukmccall on vendoring. This change only introduces a new package and adds it to 'bare-expo' behind a feature flag, hoping to not affect anyone's workflow or any existing packages.

# How

Contents of this change:
- Package 'expo-development-client'
  - Contains native code on iOS and Android to initialize a 'development client controller' module, which initializes a launcher bridge before the app's main bridge and loads the launcher's UI using an embedded JS bundle.
  - The development client controller presents a QR code scanner to load the user's app bundle from a React Native packager server.
  - The development client native code also adds extensions to the dev menu: currently just one 'Back to launcher' button when the user is in their app to go back to the initial launcher screen.
- Test in 'bare-expo'
  - On both iOS and Android, the development client package was added as expected for a regular user of the package to add.
  - Its use is kept behind a feature flag (`useDevClient` in 'AppDelegate.swift' for iOS, `USE_DEV_CLIENT` in 'MainApplication.java' for Android) so that normal 'bare-expo' workflow is preserved.

# Test Plan

- Enable the feature flags mentioned above in 'bare-expo'.
- Build and run a development build (both iOS and Android) (you don't need to run a packager in 'apps/bare-expo' yet).
  - On launch, should show the launcher screen.
- Run `EXPO_TARGET=bare expo start` in 'apps/bare-expo'.
- Scan the resulting QR code in the launcher.
  - This should load the regular 'bare-expo' app. Try using it and make sure stuff works.
- Shake to show dev menu, touch 'Back to launcher' to go back to the launcher.
  - This should present the initial launcher screen of the development client again.
- Try loading the QR code again.
  - This should properly load 'bare-expo' again.
- You can also serve and load any other app that only needs some subset of the native modules present in 'bare-expo'.
